File: app/src/main/java/openfoodfacts/github/scrachx/openfood/views/adapters/CalculatedNutrimentsGridAdapter.java
package openfoodfacts.github.scrachx.openfood.views.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import org.apache.commons.lang.StringUtils;

import java.util.List;

import openfoodfacts.github.scrachx.openfood.R;
import openfoodfacts.github.scrachx.openfood.models.NutrimentListItem;

public class CalculatedNutrimentsGridAdapter extends NutrimentsGridAdapter {
    private static final int TYPE_HEADER = 0;
    private static final int TYPE_ITEM = 1;
    private final List<NutrimentListItem> nutrimentListItems;

    public CalculatedNutrimentsGridAdapter(List<NutrimentListItem> nutrimentListItems) {
        super(nutrimentListItems);
        this.nutrimentListItems = nutrimentListItems;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        boolean isViewTypeHeader = viewType == TYPE_HEADER;

        int layoutResourceId = isViewTypeHeader ? R.layout.nutrition_fact_header_calc : R.layout.nutriment_item_list;
        View v = LayoutInflater.from(parent.getContext()).inflate(layoutResourceId, parent, false);

        if (isViewTypeHeader) {
            boolean displayServing = false;
            for (NutrimentListItem nutriment : nutrimentListItems) {
                final CharSequence servingValue = nutriment.getServingValue();
                if (servingValue != null && !StringUtils.isBlank(servingValue.toString())) {
                    displayServing = true;
                }
            }
            return new NutrimentHeaderViewHolder(v, displayServing);
        } else {
            return new NutrimentListViewHolder(v);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (!(holder instanceof NutrimentListViewHolder)) {
            return;
        }

        NutrimentListItem item = nutrimentListItems.get(position);
        NutrimentListViewHolder nutrimentListViewHolder = (NutrimentListViewHolder) holder;
        nutrimentListViewHolder.fillNutrimentValue(item);
        nutrimentListViewHolder.fillServingValue(item);
        holder.setIsRecyclable(false);
    }

}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/adapters/CalculatedNutrimentsGridAdapter.kt
package openfoodfacts.github.scrachx.openfood.features.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import openfoodfacts.github.scrachx.openfood.R
import openfoodfacts.github.scrachx.openfood.models.NutrimentListItem

class CalculatedNutrimentsGridAdapter(private val nutrimentListItems: List<NutrimentListItem>) : NutrimentsGridAdapter(nutrimentListItems) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NutrimentViewHolder {
        val isViewTypeHeader = viewType == TYPE_HEADER
        val layoutResourceId = if (isViewTypeHeader) R.layout.nutrition_fact_header_calc else R.layout.nutriment_item_list
        val view = LayoutInflater.from(parent.context).inflate(layoutResourceId, parent, false)

        return if (isViewTypeHeader) {
            val displayServing = nutrimentListItems.any { !it.servingValueStr.isNullOrBlank() }
            NutrimentViewHolder.NutrimentHeaderViewHolder(view, displayServing)
        } else {
            NutrimentViewHolder.NutrimentListViewHolder(view)
        }
    }

    override fun onBindViewHolder(holder: NutrimentViewHolder, position: Int) {
        if (holder !is NutrimentViewHolder.NutrimentListViewHolder) return

        val item = nutrimentListItems[position]
        holder.fillNutriment(item)
        holder.setIsRecyclable(false)
    }

    companion object {
        private const val TYPE_HEADER = 0
        private const val TYPE_ITEM = 1
    }
}